tango/integrations/beaker/step_cache.py

```diff
@@ -0,0 +1,178 @@
+
+        # Next check our local on-disk cache.
+        with self._acquire_step_lock_file(step, read_only_ok=True):
+            if self.step_dir(step).is_dir():
+                return load_and_return()
+
+        # Finally, check Beaker for the corresponding dataset.
+        with self._acquire_step_lock_file(step):
+            # Make sure the step wasn't cached since the last time we checked (above).
+            if self.step_dir(step).is_dir():
+                return load_and_return()
+
+            dataset = self._step_result_dataset(step)
+            if dataset is None:
+                raise KeyError(step)
+
+            # We'll download the dataset to a temporary directory first, in case something goes wrong.
+            temp_dir = tempfile.mkdtemp(dir=self.dir, prefix=key)
+            try:
+                self.beaker.dataset.fetch(dataset, target=temp_dir, quiet=True)
+                # Download and extraction was successful, rename temp directory to final step result directory.
+                os.replace(temp_dir, self.step_dir(step))
+            finally:
+                shutil.rmtree(temp_dir, ignore_errors=True)
+
+            return load_and_return()
+
+    def __setitem__(self, step: Step, value: Any) -> None:
+        if not step.cache_results:
+            logger.warning("Tried to cache step %s despite being marked as uncacheable.", step.name)
+            return
+
+        with self._acquire_step_lock_file(step):
+            # We'll write the step's results to temporary directory first, and try to upload to
+            # Beaker from there in case anything goes wrong.
+            temp_dir = Path(tempfile.mkdtemp(dir=self.dir, prefix=step.unique_id))
+            (temp_dir / Constants.STEP_RESULT_DIR).mkdir()
+            try:
+                step.format.write(value, temp_dir / Constants.STEP_RESULT_DIR)
+                metadata = CacheMetadata(step=step.unique_id, format=step.format)
+                metadata.to_params().to_file(temp_dir / self.METADATA_FILE_NAME)
+                # Create the dataset and upload serialized result to it.
+                self._sync_step_dataset(step, temp_dir)
+                # Upload successful, rename temp directory to the final step result directory.
+                if self.step_dir(step).is_dir():
+                    shutil.rmtree(self.step_dir(step), ignore_errors=True)
+                os.replace(temp_dir, self.step_dir(step))
+            finally:
+                shutil.rmtree(temp_dir, ignore_errors=True)
+
+        # Finally, add to in-memory caches.
+        self._add_to_cache(step.unique_id, value)
+
+    def __len__(self) -> int:
+        return len(
+            [
+                ds
+                for ds in self.beaker.workspace.datasets(
+                    uncommitted=False, match=Constants.STEP_DATASET_PREFIX
+                )
+                if ds.name is not None and ds.name.startswith(Constants.STEP_DATASET_PREFIX)
+            ]
+        )
```

Comment:
```suggestion
        return sum(
                1
                for ds in self.beaker.workspace.datasets(
                    uncommitted=False, match=Constants.STEP_DATASET_PREFIX
                )
                if ds.name is not None and ds.name.startswith(Constants.STEP_DATASET_PREFIX)
        )
```

This saves some memory allocations because it never creates a temporary list.